nestkernel/eprop_archiving_node.cpp

```diff
@@ -190,12 +192,21 @@ nest::EpropArchivingNode::erase_unneeded_update_history()
 void
 nest::EpropArchivingNode::erase_unneeded_eprop_history()
 {
+  if ( eprop_history_.empty()  // nothing to remove
+    or update_history_.empty() // no time markers to check
+  )
+  {
+    return;
+  }
+
   const long update_interval = kernel().simulation_manager.get_eprop_update_interval().get_steps();
 
   auto it_update_hist = update_history_.begin();
-  long t = update_history_.begin()->t_;
 
-  for ( ; t <= ( update_history_.end() - 1 )->t_ and it_update_hist != update_history_.end(); t += update_interval )
+
+  for ( long t = update_history_.begin()->t_;
+        t <= ( update_history_.end() - 1 )->t_ and it_update_hist != update_history_.end();
```

Comment:
```suggestion
        t <= ( update_history_.end() - 1 )->t_ ;
```
I added that part instead of the guard for the `update_history_` , but since it apparently does not fulfill the desired function, it can go.

Reply:
Since we are looping over `update_history_` here, I would definitely keep the check, if only to avoid that we fall off the end of `update_history_`. If, on the other hand, that is implicitly guaranteed by the check on `t`, one could remove the check from the for-loop header and instead have inside the loop
```c++
    assert( it_update_hist != update_history_.end() );
    if ( it_update_hist->t_ == t )
    {
      ++it_update_hist;
    }
```

That would more clearly communicate that we always stay inside the history.

I think it could also be helpful if you could add a comment at the beginning of the code in this method which explains how this method works. If I see it right, it looks at update times in the past and then finds corresponding sections in eprop_history, which it removes, doesn't it?

Reply:
`update_history_` contains a sorted list of unique update times, i.e., if the `update_interval` is 1000 steps, then, e.g., `[1000, 2000, 5000]` - note that some entries are potentially missing (by design). We increase from the first entry in `update_history`, e.g., 1000 steps, up to the penultimate value, i.e., in this case 2000 steps. So, it is implicitly guaranteed by the check on `t` that we do not fall off the end of `update_history_`. So, if I understand it correctly, the `assert` is not necessary.

Your understanding of the mechanism is correct. Adding doxygen comments are on my todo list - I wanted to leave this to the end, so that the code is finalized, the comments do not differ from the implementatio,  and are consistent in terms of content, style and information level.

Reply:
Assertions are never necessary, the point is just to protect us and the user against errors occuring elsewhere in the code which could lead to undefined behaviour of the program when reading beyond the end of `update_history_`. Such errors can be extremely difficult to track down, and the `assert()` has minimal overhead and can be compiled out if desired. So I think it is better to be safe than sorry.

Concerning documentation, documentation on how the code here works should go into a normal comment here in the method, the doxygen comments (in the h file) should document the class interface, including private members.